2024/08/migration-from-struts1-to-struts2.md

```diff
@@ -0,0 +1,174 @@
+        ActionMapping mapping,
+        ActionForm form,
+        HttpServletRequest request,
+        HttpServletResponse response) throws Exception {
+        // add logic
+        return (mapping.findForward("success"));
+    }
+}
+```
+
+Struts 1 actions are singleton and extends base `Action` class. Alternatively `DispatchAction` can be used but the entrry point to an Action class is `execute` method for both implementation. Actions has to be thread-safe and hence all the needed variables are handled in method scope. Finally an action methods return `ActionForward` response.
+
+On the other hand, a Struts 2 a basic action class structure would be:
+
+```java
+public class MyAction extends ActionSupport {
+   public String execute() throws Exception {
+        // add logic
+        return "success";
+   }
+}
+```
+
+A Struts 2 action class not necessarily but usually extends `ActionSupport` and usual entry point is again `execute()` method without any parameters. In Struts 2 action classes are not singleton and creates instance for each request, so Class scope variables can be used. Struts 2 finally injects HttpServletRequest into ServletRequestAware actions to complete the request flow.
+
+#### Action Configuration
+
+In Struts 1 applications action configurations are made in `struts-config.xml` located in `WEB-INF` directory. Struts 1 applications need to configure actions and action form beans in this configuration file.
+
+```XML
+<struts-config>
+
+     <form-beans>
+         <form-bean name="myForm" type="com.endpoint.blog.struts.MyForm"/>
+         ...
+     </form-beans>
+
+    <global-forwards>
+       <forward name="home" path="/home.do" redirect="true" />
+       <forward name="projects" path="/myactionProcess.do?dispatch=myDispatch" redirect="true"/>
+       ...
+   </global-forwards>
+
+   <action-mappings>
+       <action path="/myPackage/add" type="org.apache.struts.actions.ForwardAction" />
+       ...
+   </action-mappings>
+
+ </struts-config> 
+```
+
+Struts 2 configuration involves a package wrapper around the action definitions.
+
+```xml
+<struts>
+	<package name="myPackage" extends="struts-default">
+    	<action name="add" >
+        	<result>/myPackage/add.jsp</result>
+        </action>
+        ...
+    </package>
+ </struts> 
+```
+
+However, in Struts 2, actions can be defined in a more general way using Struts conventions:
+
+```xml
+<struts>
+	<constant name="struts.action.extension" value="ep"/>
+    <constant name="struts.convention.action.packages" value="com.endpoint.blog.actions.*"/>
+    ...
+</struts>
+```
+
+In this article, we covered migrating the most skeletal components from Struts 1 to Struts 2 and comparison of several features between them. Complete configuration and code migration can be done gradually from Struts 1 to Struts 2 which would involve actions, interceptors, tags and many other details. Not that straight-forward but a convention change can be seen easily.
```

Comment:
In this article we covered migrating the most basic components from Struts 1 to Struts 2 and comparison of several features between each version. Complete configuration and code migration from Struts 1 to Struts 2 can be done gradually, which would involve actions, interceptors, tags and many other details. Not that straight-forward the patterns and convention changes adopted in Struts 2 becomes clearer as you dive in.

Reply:
(Sorry, I updated my suggested change at the end here ... it caught my eye just now as I was reviewing, and I realized the last sentence was still a little awkward.)